### PR TITLE
gtrak10: working racetrack, car graphics, added test controls and coi…

### PR DESCRIPTION
**Pull request by goldnchild**
…n/start switches.

The fixfreq vsync doesn't work quite right if I specify 902 for GTRAK10_HTOTAL.  If I change the vsync_filter_timeconst formula, it will sync and you can get a full display.

### CODE REVIEW

**Comment by felipesanches** [comment]
Awesome! Thanks for working on this :-D

File: src/mame/atari/atarittl.cpp [hunk 1]
```diff
@@ -297,11 +312,21 @@ void gtrak10_state::gtrak10(machine_config &config)
 	m_video->set_vert_params( GTRAK10_VTOTAL - 8,   GTRAK10_VTOTAL - 8,       GTRAK10_VTOTAL,     GTRAK10_VTOTAL);
 	m_video->set_fieldcount(2);
```
Comment (couriersud) [inline src/mame/atari/atarittl.cpp:313]: Are you sure that the video output is interlaced? The logic in this case must generate different timings for the two fields.

**Comment by goldnchild** [comment]
> Awesome! Thanks for working on this :-D

Hi Felipe, thanks for your existing driver, it was a good model to build on and for seeing how Netlist works.

File: src/mame/atari/atarittl.cpp (hunk 1 above)
Comment (couriersud) [reply src/mame/atari/atarittl.cpp:313]: The interlace detection (fieldcount=2) I wrote from specifications if I remember it right. I really need to see the `compsync` signal and check the code against it. 
This PR is a significant step forward. I am happy to commit it once all checks have passed.

Comment (goldnchild) [reply src/mame/atari/atarittl.cpp:313]: Cool!  It's kinda fun when you can just hook things up virtually with a text description and there's a complete set of logic chips to plug in and connect like a breadboard.  Like the 555 timer, I just copied an example from JD's tank driver as a model.

**Comment by felipesanches** [comment]
Screenshot from 2022-07-01 01-42-54

Pretty cool! :-)

**Comment by goldnchild** [comment]
It's even better when it's 902 pixels wide 8-)  Just do a little hack on fixfreq.h and set the GTRAK10_HTOTAL to 902

Since there's no collision with the racetrack slowing you down, it's actually a pretty good challenge to try to finish the checkpoint cycle before the time runs out.  I think there may be just enough time to hit the checkpoints in straight lines to get to the finish line before the timer expires.  One could adjust the 555 resistances to give more or less time.

File: src/mame/atari/atarittl.cpp (hunk 1 above)
Comment (stilett0) [reply src/mame/atari/atarittl.cpp:313]: yep, @goldnchild , that's exactly how it's intended to work, just like a breadboard, or rather, like Lego! The old "external netlist as ROM" (the old "pong" as opposed to "pongf") in theory made it even more so with less compiling - in practice, ehhhh. The pain also come during development when the component you want isn't in the library of components, or if your netlist is so non-performant that you have to play around with static solvers and @couriersud 's other performance optimizing features...

**Comment by felipesanches** [comment]
Screenshot from 2022-07-01 04-06-50

yeah... but the right half of screen gets "jumpy" then.

The hack you describe is incorrect? Would that break other games? I'm not yet familiar with what's going on in that `fixfreq.h` file.

**Comment by felipesanches** [comment]
And what about those (apparently) spurious pixels on the top of the screen (as well as a smaller amount of dots at the bottom)?
Any idea about their purpose?

**Comment by goldnchild** [comment]
Those are actually the data that is used for the horizontal and vertical motion of the car.  The first 32 bytes 0x0-0x1f are the horiz and vertical motion for each of the 32 directions, and 0x20 is the "stop code" where there is no motion.

Bits 0-3 are the low nibble data for the vertical motion.
Bits 4-7 are the low nibble data for the horiz motion counter.

It doesn't hurt anything except for visual aesthetics.  They could've put more logic in to gate it off, but they were economizing.

These get duplicated at the start of each track to make them "interchangeable", but I think only the first track is actually addressed.

	// ----------- Vertical positioning of the car -----------------
	//     name,     EQ, MRQ,    S0Q,    S1Q,    S2Q,    S3Q,    D0,    D1,    D2,    D3
	TTL_9314(H7, VLd1_Q,   P, GROUND, GROUND, GROUND, GROUND, DATA4, DATA5, DATA6, DATA7)  // 4 bit latch for ROM data
	//     name,    CLK,   ENP,   ENT,     CLRQ,  LOADQ,      A,      B,      C,      D
	TTL_9316(J7, HCOUNT,     P,     P, RESET1_Q, L5_A.Q,  H7.Q0,  H7.Q1,  H7.Q2,  H7.Q3)  // bits 0-3 of vpos
	TTL_9316(K7, HCOUNT, J7.RC,     P, RESET1_Q, L5_A.Q,      P,      P,      P,      P)  // bits 4-7 of vpos
	TTL_9316(L7, HCOUNT, J7.RC, K7.RC, RESET1_Q, L5_A.Q,      P, GROUND,      P,      P)  // bits 8-11 of vpos
	TTL_7400_NAND(L5_A, J7.RC, L7.RC)
	ALIAS(1PHI2, J7.QB)
	ALIAS(1PHI3, J7.QC)
	ALIAS(1PHI4, J7.QD)

	// ------------ Horizontal positioning of the car ----------------------
	//     name,   CLK,   ENP,   ENT,      CLRQ, LOADQ,        A,      B,      C,      D
	TTL_9316(D7, CLOCK, E7.RC,     P,  RESET1_Q,  D5_B.Q,      P,      P,      P, GROUND)  // bits 4-7 of hpos
	TTL_9316(E7, CLOCK,     P,     P,  RESET1_Q,  D5_B.Q, F7_A.Q, F7_D.Q, F7_C.Q, D3_C.Q)  // bits 0-3 of hpos

	TTL_7400_NAND(D5_B, E7.RC, B6_D.Q)
	TTL_7404_INVERT(B6_D, H5_A.Q)
	TTL_7410_NAND(H5_A, D7.RC, D6_A.Q, D6_B.Q) //weird: should be D6_B.QQ according to schematics

	//        name, CLK,    J, K,     CLRQ
	TTL_74107(D6_A, D7.RC,  P, P, RESET1_Q)  // bit 8 of hpos
	TTL_74107(D6_B, D6_A.Q, P, P, RESET1_Q)  // bit 9 of hpos

	TTL_7400_NAND(D3_C, VLd1, DATA3)
	TTL_7408_AND(F7_C,  VLd1, DATA2)
	TTL_7408_AND(F7_D,  VLd1, DATA1)
	TTL_7408_AND(F7_A,  VLd1, DATA0)


Take a look at direction_coordinates.pdf, I found that very helpful.  (Super good reads at Ed Fries' blog)

https://edfries.wordpress.com/2017/10/23/finding-the-first-videogame-rom/
https://edfries.wordpress.com/2017/06/14/fixing-gran-trak-10/

What followed was a series of emails with [amazing](https://edfries.files.wordpress.com/2017/06/gt10_rom_memory_map.pdf) [things](https://edfries.files.wordpress.com/2017/06/hybrid_schematics1.pdf) [attached](https://edfries.files.wordpress.com/2017/06/direction_coordinates.pdf).

**Comment by goldnchild** [comment]
> Screenshot from 2022-07-01 04-06-50
> 
> yeah... but the right half of screen gets "jumpy" then.

I think that the original game had some vertical jitteriness too, I saw this on youtube:

https://www.youtube.com/watch?v=R014taK2pZM

Note that the left and right halves are shifting up and down alternating with each other.

> 
> The hack you describe is incorrect? Would that break other games? I'm not yet familiar with what's going on in that `fixfreq.h` file.

Yes, there's a lot going on in fixfreq that I don't understand, but there's a filter that detects the vsync and it uses a time constant.  I found that changing that time constant would enable the sync to be detected.  Otherwise you only see a single frame and then no more 8-)

It seemed to work ok for tank but I didn't test everything.

File: src/mame/atari/atarittl.cpp (hunk 1 above)
Comment (goldnchild) [reply src/mame/atari/atarittl.cpp:313]: I was thinking of a hybrid approach where the sound can be done with netlist and the video generation can be done with "regular mame code".  My computer isn't fast enough for netlist to work in the multi megahertz range at full speed, but I reckon sound is at least an order of magnitude less demanding.  (khz vs megahertz)

**Pull request merged**